eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -144,66 +161,37 @@ class NodeRelay private(nodeParams: NodeParams,
-    context.messageAdapter[MultiPartPaymentFSM.MultiPartPaymentSucceeded](WrappedMultiPartPaymentSucceeded)
-  }.toClassic
-  private val payFsmAdapters = {
-    context.messageAdapter[PreimageReceived](WrappedPreimageReceived)
-    context.messageAdapter[PaymentSent](WrappedPaymentSent)
-    context.messageAdapter[PaymentFailed](WrappedPaymentFailed)
-  }.toClassic
-
-  def apply(): Behavior[Command] =
-    Behaviors.receiveMessagePartial {
-      // We make sure we receive all payment parts before forwarding to the next trampoline node.
-      case Relay(IncomingPacket.NodeRelayPacket(add, outer, inner, next)) => outer.paymentSecret match {
-        case None =>
-          // TODO: @pm: maybe those checks should be done later in the flow (by the mpp FSM?)
-          context.log.warn("rejecting htlcId={}: missing payment secret", add.id)
-          rejectHtlc(add.id, add.channelId, add.amountMsat)
-          stopping()
-        case Some(secret) =>
-          import akka.actor.typed.scaladsl.adapter._
-          context.log.info("relaying payment relayId={}", relayId)
-          val mppFsm = context.actorOf(MultiPartPaymentFSM.props(nodeParams, add.paymentHash, outer.totalAmount, mppFsmAdapters))
-          context.log.debug("forwarding incoming htlc to the payment FSM")
-          mppFsm ! MultiPartPaymentFSM.HtlcPart(outer.totalAmount, add)
-          receiving(Queue(add), secret, inner, next, mppFsm)
-      }
-    }
-
   /**
    * We start by aggregating an incoming HTLC set. Once we received the whole set, we will compute a route to the next
    * trampoline node and forward the payment.
    *
    * @param htlcs       received incoming HTLCs for this set.
-   * @param secret      all incoming HTLCs in this set must have the same secret to protect against probing / fee theft.
    * @param nextPayload relay instructions (should be identical across HTLCs in this set).
    * @param nextPacket  trampoline onion to relay to the next trampoline node.
    * @param handler     actor handling the aggregation of the incoming HTLC set.
    */
-  private def receiving(htlcs: Queue[UpdateAddHtlc], secret: ByteVector32, nextPayload: Onion.NodeRelayPayload, nextPacket: OnionRoutingPacket, handler: ActorRef): Behavior[Command] =
+  private def receiving(htlcs: Queue[UpdateAddHtlc], nextPayload: Onion.NodeRelayPayload, nextPacket: OnionRoutingPacket, handler: ActorRef): Behavior[Command] =
     Behaviors.receiveMessagePartial {
       case Relay(IncomingPacket.NodeRelayPacket(add, outer, _, _)) => outer.paymentSecret match {
+        // TODO: @pm: maybe those checks should be done by the mpp FSM?
         case None =>
-          context.log.warn("rejecting htlcId={}: missing payment secret", add.id)
+          context.log.warn("rejecting htlc #{} from channel {}: missing payment secret", add.id, add.channelId)
           rejectHtlc(add.id, add.channelId, add.amountMsat)
           Behaviors.same
```

Comment:
Why keep this? We know it can't happen right?

Reply:
Yes, but at the type level it could (if we change the `NodeRelayer` code without updating the `NodeRelay`), so it's kind of a defense-in-depth to keep this handler (for now, let's get rid of all of them soon!)

Reply:
Another way to deal with that is to add a `require` when instantiating the actor, and then do `outer.paymentSecret.get` everywhere?

Reply:
I'll try that

Reply:
I'm not a big fan, it's less robust since we can cleanly handle this case in just a few lines of code without any risk of crashing the actor. I'd rather wait to make `payment_secret` mandatory to simplify this.

Reply:
True, it's a bit dirty.